File: 2026-03-March-GFG-POTD/March 25 - Minimum height roots/Solution.cpp
class Solution {
public:
    vector<int> minHeightRoot(int V, vector<vector<int>>& edges) {
        vector<int> result;

        // Edge case
        if (V == 1) {
            result.push_back(0);
            return result;
        }

        // Step 1: Build graph
        vector<vector<int>> adj(V);
        vector<int> degree(V, 0);

        for (auto& edge : edges) {
            int u = edge[0];
            int v = edge[1];

            adj[u].push_back(v);
            adj[v].push_back(u);

            degree[u]++;
            degree[v]++;
        }

        // Step 2: Add initial leaves
        queue<int> q;
        for (int i = 0; i < V; i++) {
            if (degree[i] == 1) {
                q.push(i);
            }
        }

        // Step 3: Trim leaves
        int remainingNodes = V;

        while (remainingNodes > 2) {
            int size = q.size();
            remainingNodes -= size;

            for (int i = 0; i < size; i++) {
                int node = q.front();
                q.pop();

                for (int neighbor : adj[node]) {
                    degree[neighbor]--;

                    if (degree[neighbor] == 1) {
                        q.push(neighbor);
                    }
                }
            }
        }

        // Remaining nodes are centers
        while (!q.empty()) {
            result.push_back(q.front());
            q.pop();
        }

        return result;
    }
};

File: 2026-03-March-GFG-POTD/March 25 - Minimum height roots/Solution.py
from collections import deque

class Solution:
    def minHeightRoot(self, V, edges):
        result = []

        # Edge case
        if V == 1:
            return [0]

        # Step 1: Build graph
        adj = [[] for _ in range(V)]
        degree = [0] * V

        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)
            degree[u] += 1
            degree[v] += 1

        # Step 2: Add initial leaves
        q = deque()
        for i in range(V):
            if degree[i] == 1:
                q.append(i)

        # Step 3: Trim leaves
        remaining_nodes = V

        while remaining_nodes > 2:
            size = len(q)
            remaining_nodes -= size

            for _ in range(size):
                node = q.popleft()

                for neighbor in adj[node]:
                    degree[neighbor] -= 1

                    if degree[neighbor] == 1:
                        q.append(neighbor)

        # Remaining nodes are centers
        return list(q)
